File: src/convert.js
/**
 * HTML -> Markdown, written against the markup Hexo and the NexT theme emit.
 *
 * Generic HTML-to-Markdown converters mangle exactly the parts a Hexo site is
 * made of: code blocks are a <table> with a line-number gutter, every heading
 * carries an anchor <a>, headings and hard breaks appear inside list items.
 * This converter knows those shapes and escapes only what would change meaning
 * in Markdown, because over-escaped Markdown is unpleasant to keep editing --
 * and the whole point of recovering the source is to edit it.
 *
 * Every rule here was earned against a real 25-post site, rendered back
 * through Hexo and diffed against the original HTML until every article body
 * matched. Comments say which rule fixed which failure. The Python
 * implementation in python/ is the reference; scripts/parity.js checks the two
 * produce byte-identical Markdown.
 *
 * Nodes are domhandler nodes as produced by cheerio: {type, name, attribs,
 * children, prev, next, data}.
 */

const HEADINGS = new Set(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']);

// Characters that are literal in the rendered HTML but would be markup in
// Markdown. No `~`: old marked does not treat `\~` as an escape and prints the
// backslash; new marked turns a pair into strikethrough. It becomes an entity.
const INLINE_ESC = /([*_`\\[\]])/g;
const LINE_START_ESC = /^(\s*)(#{1,6}|[>+-]|\d+\.)(\s)/m;

const isText = (n) => n && n.type === 'text';
const isTag = (n) => n && n.type === 'tag';
const classes = (n) => (n.attribs && n.attribs.class ? n.attribs.class.split(/\s+/).filter(Boolean) : []);
const attr = (n, k) => (n.attribs && n.attribs[k]) || '';

/** All descendant text, concatenated (cheerio decodes entities once). */
export function getText(n) {
  if (isText(n)) return n.data;
  if (!n.children) return '';
  let s = '';
  for (const c of n.children) s += getText(c);
  return s;
}

function findFirst(n, pred) {
  if (!n.children) return null;
  for (const c of n.children) {
    if (pred(c)) return c;
    const r = findFirst(c, pred);
    if (r) return r;
  }
  return null;
}

function findAll(n, pred, out = []) {
  if (!n.children) return out;
  for (const c of n.children) {
    if (pred(c)) out.push(c);
    findAll(c, pred, out);
  }
  return out;
}

const isPunct = (ch) => !!ch && (/[!"#$%&'()*+,\-./:;<=>?@[\\\]^_`{|}~]/.test(ch) || /\p{P}/u.test(ch));

export class Converter {
  convert(node) {
    let out = '';
    for (const c of node.children || []) out += this.block(c);
    out = out.replace(/\n{3,}/g, '\n\n');
    // Strip newlines only: a document that opens with an indented code block
    // must keep the four spaces on its first line.
    return out.replace(/^\n+|\n+$/g, '') + '\n';
  }

  // ------------------------------------------------------------------ blocks
  block(n) {
    if (isText(n)) {
      // Bare text at block level: hexo-renderer-marked emits this when a
      // heading is followed by text without a blank line in the source. It is
      // a paragraph in every sense but the tag.
      const t = this.guardLineStart(this.text(n).trim());
      return t ? t + '\n\n' : '';
    }
    if (!isTag(n)) return '';
    const name = n.name;
    if (HEADINGS.has(name)) return '#'.repeat(+name[1]) + ' ' + this.inlineChildren(n).trim() + '\n\n';
    if (name === 'p') {
      const body = this.guardLineStart(this.inlineChildren(n).trim());
      return body ? body + '\n\n' : '';
    }
    if (name === 'figure' && classes(n).includes('highlight')) return this.codeFigure(n);
    if (name === 'pre') {
      // A bare <pre><code> with no highlight figure around it is what marked
      // emits for an INDENTED code block; Hexo only runs the highlighter on
      // fenced ones. Re-emit it indented so the re-render is a bare <pre> too.
      const code = getText(n).replace(/\n+$/, '');
      return code.split('\n').map((ln) => (ln ? '    ' + ln : '')).join('\n') + '\n\n';
    }
    if (name === 'ul' || name === 'ol') return this.list(n, name === 'ol', 0) + '\n';
    if (name === 'blockquote') {
      let inner = '';
      for (const c of n.children) inner += this.block(c);
      inner = inner.trim();
      return inner.split('\n').map((ln) => (ln ? '> ' + ln : '>')).join('\n') + '\n\n';
    }
    if (name === 'table') return this.table(n);
    if (name === 'hr') return '---\n\n';
    if (name === 'div') {
      let s = '';
      for (const c of n.children) s += this.block(c);
      return s;
    }
    if (name === 'img') return this.img(n) + '\n\n';
    if (name === 'br') return '\n';
    if (name === 'script' || name === 'style') return '';
    const t = this.inline(n).trim();
    return t ? t + '\n\n' : '';
  }

  codeFigure(fig) {
    let lang = '';
    for (const c of classes(fig)) if (c !== 'highlight') lang = c;
    if (lang === 'plain') lang = '';
    const caption = findFirst(fig, (c) => isTag(c) && c.name === 'figcaption');
    const cap = caption ? getText(caption).replace(/\s+/g, ' ').trim() : '';
    const codeTd = findFirst(fig, (c) => isTag(c) && c.name === 'td' && classes(c).includes('code'));
    let code;
    if (!codeTd) {
      const pre = findFirst(fig, (c) => isTag(c) && c.name === 'pre');
      code = pre ? getText(pre) : getText(fig);
    } else {
      const lines = findAll(codeTd, (c) => isTag(c) && c.name === 'span' && classes(c).includes('line'));
      code = lines.length ? lines.map((l) => getText(l)).join('\n') : getText(codeTd);
    }
    code = code.replace(/\n+$/, '');
    let fence = '```';
    while (code.includes(fence)) fence += '`';
    const head = fence + lang + (cap ? ' ' + cap : '');
    return `${head}\n${code}\n${fence}\n\n`;
  }

  /**
   * Each <li> becomes one or more lines. Inline runs are joined into the first
   * line; block children (headings, nested lists, code, further paragraphs) go
   * on their own lines indented under the marker, so the renderer re-parses
   * them as blocks inside the item. A heading glued onto the item text on one
   * line is just text with hashes in it.
   */
  list(lst, ordered, depth) {
    const out = [];
    let i = 0;
    for (const li of lst.children.filter((c) => isTag(c) && c.name === 'li')) {
      i += 1;
      const marker = ordered ? `${i}. ` : '- ';
      const pad = ' '.repeat(marker.length);
      const blocks = [];
      let run = [];
      const flush = () => {
        // Concatenate as the DOM had it; text nodes carry their own spaces. A
        // hard break ("  \n" from <br>) must survive, so it is protected while
        // runs of spaces are collapsed.
        let s = run.join('').split('  \n').join('\0');
        s = s.replace(/[ \t]+/g, ' ').trim().split('\0').join('  \n');
        if (s) blocks.push(s);
        run = [];
      };
      for (const c of li.children) {
        if (isTag(c) && (c.name === 'ul' || c.name === 'ol')) {
          flush();
          blocks.push(this.list(c, c.name === 'ol', 0).replace(/\n+$/, ''));
        } else if (isTag(c) && HEADINGS.has(c.name)) {
          flush();
          blocks.push('#'.repeat(+c.name[1]) + ' ' + this.inlineChildren(c).trim());
        } else if (isTag(c) && c.name === 'figure') {
          flush();
          blocks.push(this.codeFigure(c).replace(/\n+$/, ''));
        } else if (isTag(c) && c.name === 'pre') {
          flush();
          blocks.push(this.block(c).replace(/\n+$/, ''));
        } else if (isTag(c) && (c.name === 'p' || c.name === 'div')) {
          flush();
          blocks.push(this.inlineChildren(c).trim());
        } else if (isTag(c) && c.name === 'br') {
          run.push('  \n');
        } else {
          run.push(isTag(c) ? this.inline(c) : this.text(c));
        }
      }
      flush();
      if (!blocks.length) blocks.push('');
      const [first, ...rest] = blocks;
      // A <br> inside the item yields "  \n"; the continuation must be
      // indented under the marker or it ends the list.
      const firstLines = first.split('\n');
      const lines = [marker + firstLines[0], ...firstLines.slice(1).map((ln) => (ln.trim() ? pad + ln : ''))];
      for (const b of rest) for (const ln of b.split('\n')) lines.push(ln ? pad + ln : '');
      out.push(lines.join('\n'));
    }
    const body = out.join('\n');
    return depth ? this.indent(body, depth) + '\n' : body + '\n';
  }

  indent(s, depth) {
    const pad = '  '.repeat(depth);
    return s.split('\n').map((ln) => (ln ? pad + ln : ln)).join('\n');
  }

  table(t) {
    const rows = [];
    for (const tr of findAll(t, (c) => isTag(c) && c.name === 'tr')) {
      const cells = tr.children
        .filter((c) => isTag(c) && (c.name === 'th' || c.name === 'td'))
        .map((td) => this.inlineChildren(td).trim().split('|').join('\\|').split('\n').join(' '));
      rows.push(cells);
    }
    if (!rows.length) return '';
    const width = Math.max(...rows.map((r) => r.length));
    for (const r of rows) while (r.length < width) r.push('');
    const [head, ...body] = rows;
    const lines = ['| ' + head.join(' | ') + ' |', '|' + '---|'.repeat(width)];
    for (const r of body) lines.push('| ' + r.join(' | ') + ' |');
    return lines.join('\n') + '\n\n';
  }

  // ------------------------------------------------------------------ inline
  inlineChildren(n) {
    let s = '';
    for (const c of n.children || []) s += isTag(c) ? this.inline(c) : this.text(c);
    return s;
  }

  inline(n) {
    const name = n.name;
    const cls = classes(n);
    if (name === 'a') {
      if (cls.includes('headerlink')) return ''; // theme heading anchor: drop it
      const href = attr(n, 'href');
      const txt = this.inlineChildren(n);
      if (!txt.trim()) return '';
      if (href.startsWith('/tags/') || href.startsWith('/categories/')) return txt;
      return `[${txt}](${href})`;
    }
    if (name === 'img') return this.img(n);
    if (name === 'strong' || name === 'b') return this.emphasis(n, '**', 'strong');
    if (name === 'em' || name === 'i') {
      if (cls.some((c) => c.startsWith('fa'))) return ''; // icon fonts
      return this.emphasis(n, '*', 'em');
    }
    if (name === 'del' || name === 's') return '~~' + this.inlineChildren(n).trim() + '~~';
    if (name === 'code') {
      const t = getText(n);
      const longest = Math.max(0, ...(t.match(/`+/g) || []).map((m) => m.length));
      const fence = '`'.repeat(longest + 1);
      return `${fence}${t}${fence}`;
    }
    if (name === 'br') return '  \n';
    if (name === 'span' && cls.includes('line')) return getText(n);
    if (['sup', 'sub', 'kbd', 'mark'].includes(name)) return `<${name}>${this.inlineChildren(n)}</${name}>`;
    if (['figure', 'pre', 'ul', 'ol', 'table', 'blockquote', 'div', 'p'].includes(name)) return this.block(n);
    return this.inlineChildren(n);
  }

  /**
   * `**text**` when CommonMark will accept it; an HTML tag otherwise.
   *
   * The delimiter run has to be left-flanking to open and right-flanking to
   * close, and Chinese prose trips both rules constantly: 为何物**[本篇]**
   * (opener followed by punctuation but preceded by a word) and
   * **【英雄难度】**parser (closer preceded by punctuation, followed by a word).
   * Old marked ignored the rules; the current one follows them and prints the
   * asterisks. The HTML tag says the same thing in both and keeps the text
   * byte-identical, so it is used exactly when the Markdown form would not
   * render. Unicode P* counts as punctuation, so ，。（）【】 behave like ASCII.
   */
  emphasis(n, marks, tag) {
    const inner = this.inlineChildren(n).trim();
    if (!inner) return '';
    const before = isText(n.prev) ? n.prev.data.slice(-1) : '';
    const after = isText(n.next) ? n.next.data.slice(0, 1) : '';
    const first = inner[0];
    const last = inner[inner.length - 1];
    const ws = (c) => /\s/.test(c);
    const canOpen = !ws(first) && (!isPunct(first) || before === '' || ws(before) || isPunct(before));
    const canClose = !ws(last) && (!isPunct(last) || after === '' || ws(after) || isPunct(after));
    return canOpen && canClose ? `${marks}${inner}${marks}` : `<${tag}>${inner}</${tag}>`;
  }

  img(n) {
    // A bare space ends the URL in Markdown and the image renders as text.
    const src = (attr(n, 'data-src') || attr(n, 'src')).split(' ').join('%20');
    const alt = attr(n, 'alt') || attr(n, 'title');
    return `![${alt}](${src})`;
  }

  text(s) {
    let t = s.data.replace(/[ \t\r\f\v]+/g, ' ').split('\n').join(' ');
    t = t.replace(INLINE_ESC, '\\$1');
    return t.split('~').join('&#126;');
  }

  /**
   * Escape a leading heading/list marker. Only for text that will sit at the
   * start of a paragraph line: inside a heading "1. 前言" cannot be a list, and
   * escaping it there puts a literal backslash into the page.
   */
  guardLineStart(t) {
    return t.replace(LINE_START_ESC, '$1\\$2$3');
  }
}

File: src/hexo_recover/convert.py
"""HTML -> Markdown, written against the markup Hexo and the NexT theme emit.

Generic HTML-to-Markdown converters mangle exactly the parts that matter in a
Hexo site: code blocks are rendered as a <table> with a line-number gutter,
every heading carries an anchor <a>, and headings can sit inside list items.
This converter knows those shapes. It escapes only the characters that would
change meaning in Markdown, because over-escaped Markdown is unpleasant to
keep editing -- and the whole point of recovering the source is to edit it.

The rules here were each earned against a real 25-post site, rendered back
through Hexo and diffed against the original HTML until every article body
matched. Comments say which rule fixed which failure.
"""
import html
import re
import unicodedata

from bs4 import NavigableString, Tag

HEADINGS = ("h1", "h2", "h3", "h4", "h5", "h6")


class Converter:
    """Depth-first walk producing Markdown.

    Contract: block elements return text that already ends with the right
    number of newlines; inline elements return text with no trailing newline.
    Keeping that contract is what stops the output from growing stray blank
    lines.
    """

    # Characters that are literal in the rendered HTML but would be markup in
    # Markdown. No `~`: old marked does not treat `\~` as an escape and prints
    # the backslash; new marked turns a pair into strikethrough. It is handled
    # as an entity in text() instead.
    _INLINE_ESC = re.compile(r"([*_`\\\[\]])")
    _LINE_START_ESC = re.compile(r"^(\s*)(#{1,6}|[>+-]|\d+\.)(\s)", re.M)
    _PUNCT_START = re.compile(r"^[\\\[\](){}<>\"'!?.,:;#*_~`|-]")

    def convert(self, node) -> str:
        out = "".join(self.block(c) for c in node.children)
        out = re.sub(r"\n{3,}", "\n\n", out)
        # Strip newlines only: a document that opens with an indented code
        # block must keep the four spaces on its first line.
        return out.strip("\n") + "\n"

    # ------------------------------------------------------------------ blocks
    def block(self, n) -> str:
        if isinstance(n, NavigableString):
            # Bare text at block level: hexo-renderer-marked emits this when a
            # heading is followed by text without a blank line in the source.
            # It is a paragraph in every sense but the tag.
            t = self.guard_line_start(self.text(n).strip())
            return t + "\n\n" if t else ""
        if not isinstance(n, Tag):
            return ""
        name = n.name
        if name in HEADINGS:
            return "#" * int(name[1]) + " " + self.inline_children(n).strip() + "\n\n"
        if name == "p":
            body = self.guard_line_start(self.inline_children(n).strip())
            return body + "\n\n" if body else ""
        if name == "figure" and "highlight" in n.get("class", []):
            return self.code_figure(n)
        if name == "pre":
            # A bare <pre><code> with no highlight figure around it is what
            # marked emits for an INDENTED code block; Hexo only runs the
            # highlighter on fenced ones. Re-emit it indented so the re-render
            # is a bare <pre> too, not a highlight table.
            code = html.unescape(n.get_text()).rstrip("\n")
            return "\n".join("    " + ln if ln else "" for ln in code.split("\n")) + "\n\n"
        if name in ("ul", "ol"):
            return self.list(n, ordered=(name == "ol"), depth=0) + "\n"
        if name == "blockquote":
            inner = "".join(self.block(c) for c in n.children).strip()
            return "\n".join("> " + ln if ln else ">" for ln in inner.split("\n")) + "\n\n"
        if name == "table":
            return self.table(n)
        if name == "hr":
            return "---\n\n"
        if name == "div":
            return "".join(self.block(c) for c in n.children)
        if name == "img":
            return self.img(n) + "\n\n"
        if name == "br":
            return "\n"
        if name in ("script", "style"):
            return ""
        t = self.inline(n).strip()
        return t + "\n\n" if t else ""

    def code_figure(self, fig) -> str:
        lang = ""
        for c in fig.get("class", []):
            if c != "highlight":
                lang = c
        if lang == "plain":
            lang = ""
        caption = fig.find("figcaption")
        cap = caption.get_text(" ", strip=True) if caption else ""
        code_td = fig.select_one("td.code")
        if code_td is None:
            pre = fig.find("pre")
            code = pre.get_text() if pre else fig.get_text()
        else:
            lines = code_td.select("span.line")
            code = "\n".join(self.raw_text(ln) for ln in lines) if lines else code_td.get_text()
        code = code.rstrip("\n")
        fence = "```"
        while fence in code:
            fence += "`"
        head = fence + lang + (" " + cap if cap else "")
        return f"{head}\n{code}\n{fence}\n\n"

    @staticmethod
    def raw_text(node) -> str:
        # Code must not be Markdown-escaped; only entity-decoded.
        return html.unescape(node.get_text())

    def list(self, lst, ordered, depth) -> str:
        """Each <li> becomes one or more lines. Inline runs are joined into the
        first line; block children (headings, nested lists, code, further
        paragraphs) go on their own lines indented under the marker, so the
        renderer re-parses them as blocks inside the item. A heading glued
        onto the item text on one line is just text with hashes in it."""
        out = []
        i = 0
        for li in lst.find_all("li", recursive=False):
            i += 1
            marker = f"{i}. " if ordered else "- "
            pad = " " * len(marker)
            blocks, inline_run = [], []

            def flush():
                # Concatenate as the DOM had it; text nodes carry their own
                # spaces. A hard break ("  \n" from <br>) must survive, so it is
                # protected while runs of spaces are collapsed.
                s = "".join(inline_run).replace("  \n", "\0")
                s = re.sub(r"[ \t]+", " ", s).strip().replace("\0", "  \n")
                if s:
                    blocks.append(s)
                inline_run.clear()

            for c in li.children:
                if isinstance(c, Tag) and c.name in ("ul", "ol"):
                    flush()
                    blocks.append(self.list(c, ordered=(c.name == "ol"), depth=0).rstrip("\n"))
                elif isinstance(c, Tag) and c.name in HEADINGS:
                    flush()
                    blocks.append("#" * int(c.name[1]) + " " + self.inline_children(c).strip())
                elif isinstance(c, Tag) and c.name == "figure":
                    flush()
                    blocks.append(self.code_figure(c).rstrip("\n"))
                elif isinstance(c, Tag) and c.name == "pre":
                    flush()
                    blocks.append(self.block(c).rstrip("\n"))
                elif isinstance(c, Tag) and c.name in ("p", "div"):
                    flush()
                    blocks.append(self.inline_children(c).strip())
                elif isinstance(c, Tag) and c.name == "br":
                    inline_run.append("  \n")
                else:
                    inline_run.append(self.inline(c) if isinstance(c, Tag) else self.text(c))
            flush()
            if not blocks:
                blocks = [""]
            first, rest = blocks[0], blocks[1:]
            # A <br> inside the item yields "  \n"; the continuation must be
            # indented under the marker or it ends the list.
            first_lines = first.split("\n")
            lines = [marker + first_lines[0]] + [pad + ln if ln.strip() else "" for ln in first_lines[1:]]
            for b in rest:
                lines.extend(pad + ln if ln else "" for ln in b.split("\n"))
            out.append("\n".join(lines))
        body = "\n".join(out)
        return self.indent(body, depth) + "\n" if depth else body + "\n"

    @staticmethod
    def indent(s, depth) -> str:
        pad = "  " * depth
        return "\n".join(pad + ln if ln else ln for ln in s.split("\n"))

    def table(self, t) -> str:
        rows = []
        for tr in t.find_all("tr"):
            cells = [self.inline_children(td).strip().replace("|", "\\|").replace("\n", " ")
                     for td in tr.find_all(["th", "td"])]
            rows.append(cells)
        if not rows:
            return ""
        width = max(len(r) for r in rows)
        rows = [r + [""] * (width - len(r)) for r in rows]
        head, body = rows[0], rows[1:]
        lines = ["| " + " | ".join(head) + " |", "|" + "---|" * width]
        lines += ["| " + " | ".join(r) + " |" for r in body]
        return "\n".join(lines) + "\n\n"

    # ------------------------------------------------------------------ inline
    def inline_children(self, n) -> str:
        return "".join(self.inline(c) if isinstance(c, Tag) else self.text(c) for c in n.children)

    def inline(self, n) -> str:
        name = n.name
        cls = n.get("class", [])
        if name == "a":
            if "headerlink" in cls:  # theme heading anchor: drop it
                return ""
            href = n.get("href", "")
            txt = self.inline_children(n)
            if not txt.strip():
                return ""
            if href.startswith("/tags/") or href.startswith("/categories/"):
                return txt
            return f"[{txt}]({href})"
        if name == "img":
            return self.img(n)
        if name in ("strong", "b"):
            return self.emphasis(n, "**", "strong")
        if name in ("em", "i"):
            if any(c.startswith("fa") for c in cls):  # icon fonts
                return ""
            return self.emphasis(n, "*", "em")
        if name in ("del", "s"):
            return "~~" + self.inline_children(n).strip() + "~~"
        if name == "code":
            t = html.unescape(n.get_text())
            fence = "`" * (max([len(m) for m in re.findall(r"`+", t)] + [0]) + 1)
            return f"{fence}{t}{fence}"
        if name == "br":
            return "  \n"
        if name == "span" and "line" in cls:
            return self.raw_text(n)
        if name in ("sup", "sub", "kbd", "mark"):
            return f"<{name}>{self.inline_children(n)}</{name}>"
        if name in ("figure", "pre", "ul", "ol", "table", "blockquote", "div", "p"):
            return self.block(n)
        return self.inline_children(n)

    @staticmethod
    def _is_punct(ch: str) -> bool:
        # CommonMark's "punctuation character": ASCII punctuation or any Unicode
        # P* category -- which includes ，。（）【】 and friends, the characters
        # that actually surround emphasis in Chinese prose.
        return bool(ch) and (ch in "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~" or unicodedata.category(ch).startswith("P"))

    def emphasis(self, n, marks: str, tag: str) -> str:
        """`**text**` when CommonMark will accept it; an HTML tag otherwise.

        The delimiter run has to be left-flanking to open and right-flanking to
        close, and prose in Chinese trips both rules constantly: 为何物**[本篇]**
        (opener followed by punctuation but preceded by a word), and
        **加粗**（注） (closer followed by a word... or preceded by punctuation
        and followed by more). The old marked ignored the rules; the current one
        follows them and prints the asterisks. The HTML tag says the same thing
        in both and keeps the text byte-identical, so it is used exactly when the
        Markdown form would not render."""
        inner = self.inline_children(n).strip()
        if not inner:
            return ""
        prev, nxt = n.previous_sibling, n.next_sibling
        before = str(prev)[-1:] if isinstance(prev, NavigableString) else ""
        after = str(nxt)[:1] if isinstance(nxt, NavigableString) else ""
        first, last = inner[0], inner[-1]
        # left-flanking: not followed by whitespace, and (not followed by
        # punctuation, or preceded by whitespace/punctuation/start)
        can_open = not first.isspace() and (not self._is_punct(first) or before == "" or before.isspace() or self._is_punct(before))
        # right-flanking: not preceded by whitespace, and (not preceded by
        # punctuation, or followed by whitespace/punctuation/end)
        can_close = not last.isspace() and (not self._is_punct(last) or after == "" or after.isspace() or self._is_punct(after))
        if can_open and can_close:
            return f"{marks}{inner}{marks}"
        return f"<{tag}>{inner}</{tag}>"

    @staticmethod
    def img(n) -> str:
        src = n.get("data-src") or n.get("src", "")
        # A bare space ends the URL in Markdown and the image renders as text.
        src = src.replace(" ", "%20")
        alt = n.get("alt", "") or n.get("title", "")
        return f"![{alt}]({src})"

    def text(self, s) -> str:
        t = html.unescape(str(s))
        t = re.sub(r"[ \t\r\f\v]+", " ", t).replace("\n", " ")
        t = self._INLINE_ESC.sub(r"\\\1", t)
        return t.replace("~", "&#126;")

    def guard_line_start(self, t: str) -> str:
        """Escape a leading heading/list marker. Only for text that will sit at
        the start of a paragraph line: inside a heading "1. 前言" cannot be a
        list, and escaping it there puts a literal backslash into the page."""
        return self._LINE_START_ESC.sub(r"\1\\\2\3", t, count=1)
